transport-native-epoll/src/main/c/netty_epoll_native.c

```diff
@@ -404,92 +406,221 @@ static jint netty_epoll_native_tcpMd5SigMaxKeyLen(JNIEnv* env, jclass clazz) {
 
 //LibAIO methods
 
-static jlong netty_epoll_native_createAIOContext0(JNIEnv* env, jclass clazz, jint concurrency)
+// the maximum number of buffers for a vectored IO request
+#define MAX_NUM_BUFFERS_PER_REQUEST 8
+
+typedef struct netty_iocb
 {
-    io_context_t *ctx = malloc(sizeof(io_context_t)); //Created by io_setup
-    memset(ctx, 0, sizeof(ctx));
+    struct iocb iocb;
+    int slot; // -1 when not in use, the index in the array in netty_io_context when in use
+    struct iovec iovec[MAX_NUM_BUFFERS_PER_REQUEST]; // for vectored requests, unused otherwise
+} netty_iocb_t;
+
+typedef struct netty_io_context
+{
+    io_context_t aio;
+    int concurrency;
+    netty_iocb_t* requests;
+} netty_io_context_t;
+
 
-    int r = io_setup(concurrency, ctx);
+static jlong netty_epoll_native_createAIOContext0(JNIEnv* env, jclass clazz, jint concurrency) {
+
```

Comment:
Kind of a nit, but in general we should be much more defensive in c/jni.  Like checking concurrency isn't <= 0 or > X
Also that calloc and malloc return successfully.

Reply:
Reason being we don't want a segfault

Reply:
Done. I've set a limit of 1024 since it must fit on the stack. We can revisit this in DB-1463, if we have fewer contexts and a larger concurrency then we should go back to dynamic allocation.